fix(voice): raise FileNotFoundError for missing local path in `nextcord.FFmpegPCMAudio`

Dking08 opened the pull request:
**Summary**

Detect and surface ffmpeg failures via exit code checking.

* **Affects:** `FFmpegPCMAudio`, `FFmpegOpusAudio` (via FFmpegAudio base), `AudioPlayer`
* **Changes:**
   * Added `_raise_if_process_failed()` method to check ffmpeg exit codes
   * Call error check in `FFmpeg*Audio.read()` when stream ends unexpectedly
   * Added `AudioPlayer.error` and `AudioPlayer.has_failed` properties for error inspection
   * Updated `FFmpegPCMAudio` docstring to document runtime error behavior

**Behavioral impact:**
* Previously, ffmpeg failures (including missing files) would silently result in no audio playback
* Now, ffmpeg failures raise `ClientException` with exit code information during playback
* Users can check `AudioPlayer.has_failed` and `AudioPlayer.error` to inspect playback failures

**If applicable, link issue(s):**
* Fixes: #1229

---

### ~~Initial Approach~~ (Revised - see below)

<details>
<summary>Click to view original approach (replaced with more robust solution)</summary>

**Original approach:** Proactive file existence checking

The initial implementation added early file existence validation in `FFmpegAudio.__init__` using a `_looks_like_local_path()` helper to distinguish local files from URLs/virtual inputs.

**Why it was changed:**
- Flaky and unreliable (could misidentify URLs or ffmpeg virtual inputs)
- Only caught one type of failure (missing files)
- Required complex path validation heuristics

**Revised approach:** Reactive error detection via exit codes
- More robust - catches ALL ffmpeg failures, not just missing files
- Simpler code - no complex path heuristics needed
- Better user experience - clear error messages with exit codes
- Follows established patterns - error detection at point of failure

</details>

---

**Reproduction and verification**

Minimal example:
```python
import nextcord

def after_playback(error):
    if error:
        print(f"Playback failed: {error}")

# Missing file will cause ffmpeg to exit with error
source = nextcord.FFmpegPCMAudio("non-existent-file.m4a")
voice_client.play(source, after=after_playback)

# Or check during/after playback:
if voice_client._player.has_failed:
    print(f"Error: {voice_client._player.error}")
```

**Expected:** `ClientException` raised during `.read()` when ffmpeg exits with non-zero status, error accessible via AudioPlayer properties.

**Additional notes:**
* This approach is more robust than path validation - it catches all ffmpeg failures, not just missing files
* Error detection happens at the point of failure (during read) rather than construction
* Both `FFmpegPCMAudio` and `FFmpegOpusAudio` benefit from this error handling

**This is a Code Change**
* [x] I have tested my changes.
* [x] I have updated the documentation to reflect the changes.
* [x] I have run `task pyright` and fixed the relevant issues.

**Tooling status**
* Lint/format: PASS (`task lint`)
* Typecheck: PASS (`poetry run python -m pyright` → 0 errors)
* Slotscheck: PASS (`task slotscheck`)
* Syntax/build: PASS (Poetry install / compile succeeded)

**Risk assessment**

Low. Changes improve error visibility by detecting actual ffmpeg failures rather than attempting flaky proactive validation. Error information is now accessible to users via clean API.

Dking08 committed:
fix(voice): raise FileNotFoundError for missing local path in `nextcord.FFmpegPCMAudio` (#1229)

Dking08 committed:
style(player): apply Ruff SIM fixes (SIM103, SIM102)

teaishealthy commented:
works, but im not sure if this is the right approach for this, it seems flaky. it's probably a better idea to leave the task of checking if a file exists to the user. 

somewhat confusingly nothing in the codebase ever checks the exit code of ffmpeg ever. 

some thoughts:
we could check if ffmpeg hasn't exited in FFmpeg*Audio.read and raise, but this error only gets collected in AudioPlayer._call_after and is printed to stdout, so it has no way of making it to the user

we could add something like has_failed to VoiceClient and AudioPlayer so the user could actually check why the client stopped.

Dking08 commented:
@teaishealthy Thanks for the detailed feedback!
Yeah you are right, the earlier approach was kind of proactive and flawed.
I've implemented the changes you suggested - added exit code checking in read() and exposed error/has_failed properties on AudioPlayer.
Now it gives users a reliable way to detect failures without brittle path checks.

Dking08 committed:
fix(voice): surface ffmpeg failures with exit codes

- Remove unreliable file path validation
- Add _raise_if_process_failed() method to check ffmpeg exit codes
- Expose AudioPlayer.error and has_failed properties
- Raise ClientException in read() when ffmpeg exits with non-zero status

Replaces flaky proactive file checks with robust error detection.

teaishealthy commented on nextcord/player.py:
```diff
@@ -257,7 +276,8 @@ class FFmpegPCMAudio(FFmpegAudio):
     ClientException
-        The subprocess failed to be created.
+        The subprocess failed to be created. Runtime ffmpeg failures are surfaced during
+        reads via this exception if the underlying process exits with a non-zero status.
```
```suggestion
        The subprocess failed to be created.
```

teaishealthy commented on nextcord/player.py:
```diff
@@ -791,6 +816,16 @@ def _set_source(self, source: AudioSource) -> None:
+    def has_failed(self) -> bool:
+        """bool: Whether playback terminated due to an exception."""
+        return self._current_error is not None
+
```
Add these in VoiceClient instead, the user has no good way of accessing them,  they also won't be included in the documentation build.

Dking08 replied:
Thanks for the review! I’ve added the has_failed and related error methods to the VoiceClient as suggested. Looking forward to your feedback.

Dking08 committed:
Update nextcord/player.py

Co-authored-by: teaishealthy <76410798+teaishealthy@users.noreply.github.com>

Dking08 committed:
fix(voice): handle player errors gracefully and expose last error state

Dking08 committed:
fix(voice): fixed pyright error

Dking08 commented:
@teaishealthy
Im looking forward to your feedback on this PR. If everything looks good and no further changes are needed, could you please atleast label it as `hacktoberfest-accepted`? It would mean a lot and really help me out.

Dking08 committed:
Merge branch 'master' into fix/voice-ffmpeg-filenotfound

Dking08 commented:
I am not sure where the changes are requested @ooliver1 

"Merging blocked: 1 review requesting changes and 1 approving review by reviewers with write access." - what do I need to do?
